pingcap/tiflash
# Add timeout for wait index

### What problem does this PR solve?

Issue Number: close #825

Problem Summary: Add timeout for wait index, return Region unavailable under extreme case. And it may helpful for breaking potential deadlocks between Raft log applying and coprocessor reading.

### What is changed and how it works?

* Add `flash.wait_index_timeout_ms` to set the wait index timeout
  * `0` means wait infinitely
  * Else wait index with a timeout
* If timeout happened
  * A warning log with " wait learner index {index} timeout"
  * A profile event "RaftWaitIndexTimeout" is reported to Grafana

### Related changes

- Need to cherry-pick to the release branch:

### Check List 

Tests 


- Manual test (add detailed scripts or steps below)
  - Use failpoint to pause apply raft log and make wait index timeout
  - mysql> create table t (a int); alter table t set tiflash replica 2;
  - ch> DBGInvoke enable_fail_point(pause_before_apply_raft_cmd);
  - mysql> insert into t values (10), (20), (30);select count(*) from t;
  - The timeout is shown in the "Raft Wait index Duration" panel
![image](https://user-images.githubusercontent.com/4865550/129711277-f1a06e68-5d07-417e-8544-76901626cf19.png)


Side effects



### Release note 

```release-note
Support setting timeout for Raft wait index
```
